### PR TITLE
drivers/dht: correct interpreting raw values

### LINKED ISSUE #16900: drivers/dht : driver not providing full accuracy
#### Description
DHT11 driver measures temperature and humidity. When using the Arduino example, the temperature is accurate to 0.1°C. However, when using the RIOT example (or using the driver in another application) it is only accurate to 1°C. 

Arduino:
```
23:08:38.844 -> Humidity: 49.00%  Temperature: 26.80°C 80.24°F  Heat index: 27.18°C 80.92°F
23:08:40.863 -> Humidity: 49.00%  Temperature: 26.70°C 80.06°F  Heat index: 27.09°C 80.76°F
23:08:42.915 -> Humidity: 49.00%  Temperature: 26.70°C 80.06°F  Heat index: 27.09°C 80.76°F
23:08:44.934 -> Humidity: 49.00%  Temperature: 26.70°C 80.06°F  Heat index: 27.09°C 80.76°F
23:08:46.954 -> Humidity: 49.00%  Temperature: 26.60°C 79.88°F  Heat index: 27.00°C 80.61°F
23:08:48.973 -> Humidity: 49.00%  Temperature: 26.60°C 79.88°F  Heat index: 27.00°C 80.61°F
23:08:50.992 -> Humidity: 49.00%  Temperature: 26.60°C 79.88°F  Heat index: 27.00°C 80.61°F
23:08:53.010 -> Humidity: 49.00%  Temperature: 26.60°C 79.88°F  Heat index: 27.00°C 80.61°F
23:08:55.062 -> Humidity: 49.00%  Temperature: 26.60°C 79.88°F  Heat index: 27.00°C 80.61°F
```
RIOT:
```
2021-09-27 23:07:27,746 # read
2021-09-27 23:07:27,749 # read
2021-09-27 23:07:27,749 # read
2021-09-27 23:07:27,755 # RAW values: temp:    6401 hum:   13568
2021-09-27 23:07:27,759 # DHT values - temp: 25.0°C - relative humidity: 53.0%
2021-09-27 23:07:29,770 # read
2021-09-27 23:07:29,773 # read
2021-09-27 23:07:29,773 # read
2021-09-27 23:07:29,779 # RAW values: temp:    6401 hum:   13568
2021-09-27 23:07:29,784 # DHT values - temp: 25.0°C - relative humidity: 53.0%
2021-09-27 23:07:31,794 # read
2021-09-27 23:07:31,797 # read
2021-09-27 23:07:31,797 # read
2021-09-27 23:07:31,803 # RAW values: temp:    6401 hum:   13568
2021-09-27 23:07:31,808 # DHT values - temp: 25.0°C - relative humidity: 53.0%
2021-09-27 23:07:33,819 # read
2021-09-27 23:07:33,821 # read
2021-09-27 23:07:33,822 # read
2021-09-27 23:07:33,827 # RAW values: temp:    6401 hum:   13568
2021-09-27 23:07:33,832 # DHT values - temp: 25.0°C - relative humidity: 53.0%
2021-09-27 23:07:35,843 # read
2021-09-27 23:07:35,845 # read
2021-09-27 23:07:35,846 # read
2021-09-27 23:07:35,851 # RAW values: temp:    6401 hum:   14080
2021-09-27 23:07:35,856 # DHT values - temp: 25.0°C - relative humidity: 55.0%
2021-09-27 23:07:37,867 # read
2021-09-27 23:07:37,870 # read
2021-09-27 23:07:37,870 # read
2021-09-27 23:07:37,876 # RAW values: temp:    6401 hum:   14592
2021-09-27 23:07:37,880 # DHT values - temp: 25.0°C - relative humidity: 57.0%
2021-09-27 23:07:39,891 # read
2021-09-27 23:07:39,894 # read
2021-09-27 23:07:39,894 # read
2021-09-27 23:07:39,900 # RAW values: temp:    6401 hum:   15104
2021-09-27 23:07:39,904 # DHT values - temp: 25.0°C - relative humidity: 59.0%
2021-09-27 23:07:41,915 # read
2021-09-27 23:07:41,918 # read
2021-09-27 23:07:41,918 # read
2021-09-27 23:07:41,924 # RAW values: temp:    6402 hum:   17152
2021-09-27 23:07:41,929 # DHT values - temp: 25.0°C - relative humidity: 67.0%
2021-09-27 23:07:43,939 # read
2021-09-27 23:07:43,942 # read
2021-09-27 23:07:43,942 # read
2021-09-27 23:07:43,948 # RAW values: temp:    6403 hum:   18432
2021-09-27 23:07:43,953 # DHT values - temp: 25.0°C - relative humidity: 72.0%
2021-09-27 23:07:45,963 # read
2021-09-27 23:07:45,966 # read
2021-09-27 23:07:45,966 # read
2021-09-27 23:07:45,972 # RAW values: temp:    6404 hum:   19200
2021-09-27 23:07:45,977 # DHT values - temp: 25.0°C - relative humidity: 75.0%
2021-09-27 23:07:47,987 # read
2021-09-27 23:07:47,990 # read
2021-09-27 23:07:47,991 # read
2021-09-27 23:07:47,996 # RAW values: temp:    6406 hum:   19456
2021-09-27 23:07:48,001 # DHT values - temp: 25.0°C - relative humidity: 76.0%
2021-09-27 23:07:50,012 # read
2021-09-27 23:07:50,014 # read
2021-09-27 23:07:50,015 # read
2021-09-27 23:07:50,020 # RAW values: temp:    6408 hum:   19456
2021-09-27 23:07:50,025 # DHT values - temp: 25.0°C - relative humidity: 76.0%
2021-09-27 23:07:52,036 # read
2021-09-27 23:07:52,038 # read
2021-09-27 23:07:52,039 # read
2021-09-27 23:07:52,044 # RAW values: temp:    6656 hum:   19968
2021-09-27 23:07:52,049 # DHT values - temp: 26.0°C - relative humidity: 78.0%
2021-09-27 23:07:54,060 # read
2021-09-27 23:07:54,062 # read
2021-09-27 23:07:54,063 # read
2021-09-27 23:07:54,069 # RAW values: temp:    6659 hum:   19968
2021-09-27 23:07:54,073 # DHT values - temp: 26.0°C - relative humidity: 78.0%
```

Note that while running this test I had my finger on the sensor to slowly heat it up. What is remarkable in the logs, is that after rising gradually, it suddenly jumps from 6408 to 6656 (+200). Or in a previous test: from 6665 to 6914 (+300). 
These jumps make the measured value go from 25 to 26°C or 26 to 27°C respectively.

When checking this in binary-form it becomes noticeable that the 8 LSB barely change, but the lowest bits in the MSB shift and/or get toggled. It appears the 8 LSB are not read correctly. But I could not find why this is. 

#### Steps to reproduce the issue
Run the DHT11 example. 

#### Expected results
0.1°C accurate

#### Actual results
1°C accurate

#### Versions
```
Operating System Environment
----------------------------
         Operating System: "Arch Linux" 
                   Kernel: Linux 5.13.13-arch1-1 x86_64 unknown
             System shell: GNU bash, version 5.1.8(1)-release (x86_64-pc-linux-gnu)
             make's shell: GNU bash, version 5.1.8(1)-release (x86_64-pc-linux-gnu)

Installed compiler toolchains
-----------------------------
               native gcc: gcc (GCC) 11.1.0
        arm-none-eabi-gcc: arm-none-eabi-gcc (Arch Repository) 11.2.0
                  avr-gcc: avr-gcc (GCC) 11.2.0
         mips-mti-elf-gcc: missing
           msp430-elf-gcc: missing
       riscv-none-elf-gcc: missing
  riscv64-unknown-elf-gcc: missing
     riscv-none-embed-gcc: missing
     xtensa-esp32-elf-gcc: xtensa-esp32-elf-gcc (crosstool-NG crosstool-ng-1.22.0-80-g6c4433a5) 5.2.0
   xtensa-esp8266-elf-gcc: missing
                    clang: clang version 12.0.1

Installed compiler libs
-----------------------
     arm-none-eabi-newlib: "4.1.0"
      mips-mti-elf-newlib: missing
        msp430-elf-newlib: missing
    riscv-none-elf-newlib: missing
riscv64-unknown-elf-newlib: missing
  riscv-none-embed-newlib: missing
  xtensa-esp32-elf-newlib: "2.2.0"
xtensa-esp8266-elf-newlib: missing
                 avr-libc: "2.0.0" ("20150208")

Installed development tools
---------------------------
                   ccache: missing
                    cmake: cmake version 3.21.2
                 cppcheck: missing
                  doxygen: missing
                      git: git version 2.33.0
                     make: GNU Make 4.3
                  openocd: Open On-Chip Debugger 0.11.0
                   python: Python 3.9.6
                  python2: Python 2.7.18
                  python3: Python 3.9.6
                   flake8: error: /usr/bin/python3: No module named flake8
               coccinelle: missing
```

### PR DESCRIPTION
**Pull request by wosym**
### Contribution description

The existing driver only had an accuracy of 1°C instead of 0.1°C. This was due to an incorrect interpretation of the datasheet. The lower bits were discarded. This patch corrects the use of how the raw data should be interpreted and converted. 

The proof of the pudding is in the eating:
```
2021-10-02 00:56:24,832 # RAW values: temp: 23.6 hum: 52.0
2021-10-02 00:56:24,837 # DHT values - temp: 23.6°C - relative humidity: 52.0%
2021-10-02 00:56:26,857 # RAW values: temp: 22.9 hum: 54.0
2021-10-02 00:56:26,861 # DHT values - temp: 22.9°C - relative humidity: 54.0%
2021-10-02 00:56:28,881 # RAW values: temp: 22.9 hum: 54.0
2021-10-02 00:56:28,886 # DHT values - temp: 22.9°C - relative humidity: 54.0%
2021-10-02 00:56:30,905 # RAW values: temp: 22.9 hum: 54.0
2021-10-02 00:56:30,910 # DHT values - temp: 22.9°C - relative humidity: 54.0%
```

### Testing procedure

Run tests/drver_dht

NOTE: I only have a DHT11 available, while this driver is meant to also support DHT21 and 22. I checked the datasheet, and besides the MSB being a sign bit the implementation should be identical, but... I was not able to confirm this by testing it so if anyone else has this sensor... feel free!

### Issues/PRs references

Fixes  #16900

### CODE REVIEW

**Comment by wosym** [comment]
This is not about saving CPU cycles. Heck, I don't even know if my implementation uses more or less. 
It's because the calculations were, as far as my untrained eye could see, wrong. It appears like the original author assumed that the upper and lower register were two parts of a 16-bit integer. He then combined them and divided to a meaningful number. That's not what the datasheet describes the registers represent.

**Comment by benpicco** [comment]
It would be good to verify this still yields correct results with the old sensor - does anyone have one of them?

**Comment by fjmolinas** [comment]
Whats the status here @maribu @wosym? Should your patch be applied and we merge this one?

**Comment by wosym** [comment]
@fjmolinas Sorry, I've been a bit preoccupied lately. I'll try to have it finished up by tomorrow evening.

**Comment by wosym** [comment]
Maribu's suggestion of removing the distinction entirely was indeed a good one. I modified the code (I think I ended up with an almost identical patch to what Maribu had), and tested it again: everything still works as expected.

At first I considered removing the if/else entirely and turn it into a one-liner. It would save a couple of clockticks, but for clarity sake I think it's best to leave the if/else in there. 

If you agree, feel free to merge. If you object, let me know and I will adjust it again.

Apologies for the long delay!

(Ps. github-actions labeled this as cpu and AVR somehow? Did I touch any files that are related to AVR without realizing it?)

File: drivers/dht/dht.c [hunk 1]
```diff
@@ -178,33 +184,23 @@ int dht_read(dht_t *dev, int16_t *temp, int16_t *hum)
...
-                if (raw_temp & 0x8000) {
-                    dev->last_val.temperature = (int16_t)((raw_temp & ~0x8000) * -1);
-                }
-                else {
-                    dev->last_val.temperature = (int16_t)raw_temp;
-                }
-                break;
-            default:
-                return DHT_NODEV;      /* this should never be reached */
+        DEBUG("RAW values: temp: %2i.%i hum: %2i.%i\n", (int)raw_temp_I,
```
Comment (maribu) [inline drivers/dht/dht.c:194]:
```suggestion
        DEBUG("[dht] RAW values: temp: %2i.%i hum: %2i.%i\n", (int)raw_temp_I,
```

prefixing debug messages makes it easier to distinguish between them, when debugging issues affecting multiple modules

File: drivers/dht/dht.c [hunk 2]
```diff
@@ -126,8 +125,8 @@ int dht_init(dht_t *dev, const dht_params_t *params)
 
 int dht_read(dht_t *dev, int16_t *temp, int16_t *hum)
 {
-    uint16_t csum;
-    uint16_t raw_hum, raw_temp;
+    uint8_t csum;
+    uint8_t raw_temp_I, raw_temp_D, raw_hum_I, raw_hum_D;
```
Comment (maribu) [inline drivers/dht/dht.c:129]: Note that in C uppercase is usually exclusively used for preprocessor macros, which use all upper case

```suggestion
    uint8_t raw_temp_i, raw_temp_d, raw_hum_i, raw_hum_d;
```

File: drivers/dht/dht.c [hunk 3]
```diff
@@ -178,33 +184,23 @@ int dht_read(dht_t *dev, int16_t *temp, int16_t *hum)
...
-                break;
-            default:
-                return DHT_NODEV;      /* this should never be reached */
+        DEBUG("RAW values: temp: %2i.%i hum: %2i.%i\n", (int)raw_temp_I,
+            (int)raw_temp_D, (int)raw_hum_I, (int)raw_hum_D);
+
+        dev->last_val.humidity = raw_hum_I * 10 + raw_hum_D;
+        /* MSB set means negative temperature on DHT22. Will always be 0 on DHT11 */
+        if (raw_temp_I & 0x80) {
+            dev->last_val.temperature = ((raw_temp_I & ~0x80) * 10 + raw_temp_D) * -1;
```
Comment (maribu) [inline drivers/dht/dht.c:200]:
```suggestion
            dev->last_val.temperature = -((raw_temp_I & ~0x80) * 10 + raw_temp_D);
```

**Review by maribu — COMMENTED** [review]
Looks good to me. IMO it would make sense to split the extension of the code owner file into a separate commit (but within this PR). Squash at will.

File: drivers/dht/dht.c (hunk 1 above)
Comment (wosym) [reply drivers/dht/dht.c]: Done (Also on other DEBUGs)

**Comment by wosym** [comment]
> Looks good to me. IMO it would make sense to split the extension of the code owner file into a separate commit (but within this PR). Squash at will.

I assumed it was preferred to be done in the same commit. Good to know that this assumption was wrong! I split it off into a separate commit.

**Review by maribu — APPROVED** [review]
ACK. I did test that old DHT11 can be treated the same way as new DHT11 or DHT2*. So IMO we can just trust your testing and don't to retest again on legacy DHT11 and DHT22.

**Comment by wosym** [comment]
Thanks for the help and testing @maribu ! :)

**Comment by maribu** [comment]
Thanks for fixing the driver :-)

**Comment by hugueslarrive** [comment]
> Basically, the new DHT11 and DHT22 speak the exact same protocol with the same resolution, only the accuracy differs.

I'm not so sure about that as it breaks my end-of-2020 application that uses an AM2301 sensor (DHT21?).

For this sensor, there isn't one byte for the integral part and one for the decimal part. Instead, the two bytes form a 16-bit value (in tenths of a percent or tenths of a degree), so the high byte needs to be multiplied by 256 instead of 10 (left-shifted by 8), as explained in section 7.2 of the manual: https://www.haoyuelectronics.com/Attachment/AM2301/AM2301.pdf

Result of tests/drivers/dht:
```
2023-05-26 23:11:38,727 # DHT values - temp: 21.0°C - relative humidity: 5.9%
2023-05-26 23:11:40,753 # DHT values - temp: 21.0°C - relative humidity: 5.8%
2023-05-26 23:11:42,778 # DHT values - temp: 21.1°C - relative humidity: 6.0%
2023-05-26 23:11:44,802 # DHT values - temp: 21.0°C - relative humidity: 5.8%
2023-05-26 23:11:46,828 # DHT values - temp: 21.0°C - relative humidity: 5.9%
2023-05-26 23:11:48,853 # DHT values - temp: 21.0°C - relative humidity: 5.8%
2023-05-26 23:11:50,877 # DHT values - temp: 21.1°C - relative humidity: 6.2%
2023-05-26 23:11:52,902 # DHT values - temp: 21.0°C - relative humidity: 6.3%
2023-05-26 23:11:54,928 # DHT values - temp: 22.1°C - relative humidity: 21.4%
2023-05-26 23:11:56,952 # DHT values - temp: 23.0°C - relative humidity: 26.1%
2023-05-26 23:11:58,978 # DHT values - temp: 23.2°C - relative humidity: 26.1%
2023-05-26 23:12:01,003 # DHT values - temp: 24.8°C - relative humidity: 26.1%
2023-05-26 23:12:03,028 # DHT values - temp: 24.7°C - relative humidity: 26.1%
2023-05-26 23:12:05,053 # DHT values - temp: 25.0°C - relative humidity: 26.1%
2023-05-26 23:12:07,078 # DHT values - temp: 1.9°C - relative humidity: 26.1%
2023-05-26 23:12:09,103 # DHT values - temp: 1.6°C - relative humidity: 26.1%
2023-05-26 23:12:11,128 # DHT values - temp: 2.6°C - relative humidity: 26.1%
2023-05-26 23:12:13,153 # DHT values - temp: 2.6°C - relative humidity: 26.1%
2023-05-26 23:12:15,220 # DHT values - temp: 2.2°C - relative humidity: 26.1%
2023-05-26 23:12:17,245 # DHT values - temp: 1.9°C - relative humidity: 26.1%
2023-05-26 23:12:19,270 # DHT values - temp: 1.8°C - relative humidity: 26.1%
2023-05-26 23:12:21,295 # DHT values - temp: 1.6°C - relative humidity: 26.1%
```

**Comment by wosym** [comment]
@hugueslarrive Mhmmm... First time I hear of that sensor, but Google seems to be inconclusive about what it actually is. One source says (translated) "The AM2301 is significantly more accurate than the DHT11 and similar tot the DHT22". This would imply that it's a different sensor and would also need a different driver. (Or an extra conditional to this one)

However, other sources say that it's just a repackaged version of the DHT21, and that the other AM230x sensors are also repackaged versions of DHTxy sensors. 

I think it's going to be very hard to create a generic driver if the Chinese can't come up with a consistent numbering scheme for their products...

**Comment by hugueslarrive** [comment]
I think I order them as DHT21...

Here is the old code that differentiated between DHT11 and DHT21/22:
```c
        /* parse the RAW values */
        DEBUG("RAW values: temp: %7i hum: %7i\n", (int)raw_temp, (int)raw_hum);
        switch (dev->params.type) {
            case DHT11:
                dev->last_val.temperature = (int16_t)((raw_temp >> 8) * 10);
                dev->last_val.humidity = (int16_t)((raw_hum >> 8) * 10);
                break;
            /* DHT21 == DHT22 (same value in enum), so both are handled here */
            case DHT22:
                dev->last_val.humidity = (int16_t)raw_hum;
                /* if the high-bit is set, the value is negative */
                if (raw_temp & 0x8000) {
                    dev->last_val.temperature = (int16_t)((raw_temp & ~0x8000) * -1);
                }
                else {
                    dev->last_val.temperature = (int16_t)raw_temp;
                }
                break;
            default:
                return DHT_NODEV;      /* this should never be reached */
        }
```
As you can see, in the case of DHT11, the 8 most significant bits were right-shifted by 8 (divided by 256) and then multiplied by 10. However, for DHT21/DHT22, the 16-bit values (in tenths) were copied as is.

**Comment by wosym** [comment]
Mhmmmm, when looking back it does indeed not make much sense that the distinction was removed. The original [issue](https://github.com/RIOT-OS/RIOT/issues/16900) was also for DHT11, and not for 21/22. It was for 11 that the calculation was wrong.

@maribu do you remember why you suggested dropping the distinction? From reading the comments, I also think we never tested it (properly) on DHT21/22?

**Comment by maribu** [comment]
In https://github.com/RIOT-OS/RIOT/pull/16934#issuecomment-945173499 I suggested to drop special handling for the DHT11 case and just treat it as DHT22 as well, as there was no difference in the computation. But it seems that while there was indeed no difference in the two cases in the PR at this stage, in master prior to this PR the computation was indeed different. And now everything is treated as DHT11, so DHT2x sensor readings are now broken.

Sorry for not double checking the PR. I'll provide a PR to fix this on Tuesday.

**Comment by hugueslarrive** [comment]
I have opened PR #19674 since I had already fixed the code...

**Pull request merged**